walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
+    * events so that all attendees are free 
+    */
+  public Collection<TimeRange> queryWithoutOptionalAttendees(Collection<Event> events, MeetingRequest request) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      // if the meeting lasts more than a day, there is no solution
+      return new ArrayList<TimeRange>();
+    }
+
+    // array that will represent the number of meetings happening during that minute
+    ArrayList<Integer> meetings = new ArrayList<Integer>();
+    // before processing the events, the number of meetings for each minute is 0
+    meetings.addAll(Collections.nCopies(TimeRange.END_OF_DAY - TimeRange.START_OF_DAY + 1, 0));
+    Collection<String> attendees = request.getAttendees();
+
+    // If at least one requested attendee is participating in the event
+    // update the number of meetings for the start and end time of event
+    events.parallelStream().filter(event -> (event.containsRequestedAttendees(attendees) == true))
+                            .map(Event::getWhen)
+                            .collect(Collectors.toList())
+                            .forEach(eventWhen -> updateNumberOfMeetings(meetings, eventWhen));
+
+    // after all events are processed, for each minute x, meetings[x] will represent the number of
+    // meetings that start at minute x, minus the number of meetings that end right before minute x
+
+    // precompute the prefix sum on the meetings array s.t meetings[x] will represent the number of
+    // meetings that happen at minute x:
+    // (before computing the sum):  [ .... +1 ........ -1 .... ]
+    // (after computing the sum) :  [..... +1 +1 ... +1 0 .... ]
+    precomputePrefixSum(meetings);
+
+    return findAvailableTimeRanges(meetings, (int)request.getDuration());
+  }
+
+  /** In meetings array, add 1 to the start time of the meeting and substract 1 from the end time
+    * Only the endpoints are changed s.t. after all events are processed and the prefix sum is
+    * computed, the number of meetings increases in the array starting from start time and ending
+    * right before the end time
+   */
+  private void updateNumberOfMeetings(ArrayList<Integer> meetings, TimeRange when) {
+    // mark that a new meeting starts at when.start()
+    meetings.set(when.start(), meetings.get(when.start()) + 1); 
+    // mark that a new meeting ends right before when.end() 
+    if (when.end() <= TimeRange.END_OF_DAY) {
+      // only update the end time of events that end before the day ends
+      meetings.set(when.end(), meetings.get(when.end()) - 1);
+    }
+  }
+
+  private void precomputePrefixSum(ArrayList<Integer> meetings) {
+    // compute the sum starting from the second element because the prefix sum for the first 
+    // element is the first element 
+    for (int i = TimeRange.START_OF_DAY + 1; i <= TimeRange.END_OF_DAY; ++ i) {
+      meetings.set(i, meetings.get(i - 1) + meetings.get(i));
+    }
+  }
```

Comment:
Thank you for the idea.
 I changed the code to 
```
public Collection<TimeRange> queryWithoutOptionalAttendees(Collection<Event> events, MeetingRequest request) {
  // ...
  return findAvailableTimeRanges(precomputePrefixSum(meetings), (int)request.getDuration());
}
```
Using `meetings = precomputePrefixSum(meetings);` will result in a compilation error (meetings is used in a lambda function and if I understood it correctly, it must be final or effectively final).

Is there a better way to change the code to avoid the error?